### PR TITLE
Introduce `SaveableListener#onDeleted`

### PR DESCRIPTION
**Pull request by Vlatombe**
Usually `Saveable` objects are written, but it can happen on occasion that they get deleted, and it wasn't generating an event for every case.

This provides a more fine-grained event that can be handled by implemented listeners.

In my case, I have a use case in CloudBees CI where I need to clear a cache entry when a `Saveable` gets deleted from disk.

### Proposed changelog entries

- Developer: Add a new method `SaveableListener#onDeleted` so that plugins can be notified when a `Saveable` is deleted.

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [ ] The Jira issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [x] There is automated testing or an explanation as to why this change has no tests.
- [x] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@mention

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

File: core/src/main/java/hudson/logging/LogRecorder.java
Comment (Vlatombe) [reply core/src/main/java/hudson/logging/LogRecorder.java:559]: Yes, I guess it is unlikely enough.

File: core/src/main/java/hudson/model/AbstractItem.java
Comment (jglick) [inline core/src/main/java/hudson/model/AbstractItem.java:818]: A related oddity FTR: https://github.com/jenkinsci/jenkins/pull/9304#issuecomment-2273824789

**Review by jglick — APPROVED** [review]

File: core/src/main/java/hudson/model/listeners/SaveableListener.java
Comment (timja) [reply core/src/main/java/hudson/model/listeners/SaveableListener.java:90]: if its crucial should it be tested?

**Review by timja — APPROVED** [review]

**Review by jglick — APPROVED** [review]

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after the pending security release, we will merge it if there's no negative feedback.

Thanks!

**Pull request merged**
